cpp/src/join/mixed_join.cu

```diff
@@ -0,0 +1,1068 @@
+    switch (join_type) {
+      // Anti and semi return all the row indices from left
+      // with a corresponding NULL from the right.
+      case join_kind::LEFT_ANTI_JOIN:
+        return get_trivial_left_join_indices(left_conditional, stream).first;
+      // Inner and left semi joins return empty output because no matches can exist.
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr);
+      default: CUDF_FAIL("Invalid join kind."); break;
+    }
+  } else if (left_num_rows == 0) {
+    switch (join_type) {
+      // Anti and semi joins both return empty sets.
+      case join_kind::LEFT_ANTI_JOIN:
+      case join_kind::LEFT_SEMI_JOIN:
+        return std::make_unique<rmm::device_uvector<size_type>>(0, stream, mr);
+      default: CUDF_FAIL("Invalid join kind."); break;
+    }
+  }
+
+  // If evaluating the expression may produce null outputs we create a nullable
+  // output column and follow the null-supporting expression evaluation code
+  // path.
+  auto const has_nulls =
+    cudf::has_nulls(left_equality) || cudf::has_nulls(right_equality) ||
+    binary_predicate.may_evaluate_null(left_conditional, right_conditional, stream);
+
+  auto const parser = ast::detail::expression_parser{
+    binary_predicate, left_conditional, right_conditional, has_nulls, stream, mr};
+  CUDF_EXPECTS(parser.output_type().id() == type_id::BOOL8,
+               "The expression must produce a boolean output.");
+
+  // TODO: The non-conditional join impls start with a dictionary matching,
+  // figure out what that is and what it's needed for (and if conditional joins
+  // need to do the same).
+  auto& probe     = swap_tables ? right_equality : left_equality;
+  auto& build     = swap_tables ? left_equality : right_equality;
+  auto probe_view = table_device_view::create(probe, stream);
+  auto build_view = table_device_view::create(build, stream);
+  row_equality equality_probe{
+    cudf::nullate::DYNAMIC{has_nulls}, *probe_view, *build_view, compare_nulls};
+
+  semi_map_type hash_table{compute_hash_table_size(build.num_rows()),
+                           std::numeric_limits<hash_value_type>::max(),
+                           cudf::detail::JoinNoneValue,
+                           detail::hash_table_allocator_type{default_allocator<char>{}, stream},
+                           stream.value()};
+
+  // Create hash table containing all keys found in right table
+  // TODO: To add support for nested columns we will need to flatten in many
+  // places. However, this probably isn't worth adding any time soon since we
+  // won't be able to support AST conditions for those types anyway.
+  auto const build_nulls = cudf::nullate::DYNAMIC{cudf::has_nulls(build)};
+  row_hash const hash_build{build_nulls, *build_view};
+  // Since we may see multiple rows that are identical in the equality tables
+  // but differ in the conditional tables, we must force insertion of all the
```

Comment:
This comment, repeated elsewhere, seems misleading.  The condition tables could also be exactly identical between the relevant rows, and we still can't get away with ignoring duplicates.  For a semi join, every left table row that matches is emitted even if there are other left table rows that are identical.  So this seems more about ensuring we track every row in the left table that matched the hash keys despite the possibility of duplicate keys rather than what the contents of the condition tables could be.

Reply:
I think you're misunderstanding the purpose of this, so let me explain and maybe you can come up with a more enlightening comment. You're right that we need to emit every left row, but we build the hash table on the rows of the right table and then probe with the left rows. The problem case is, what happens if two rows in the right equality table are identical, but those rows have different contents in the right conditional tables? When we construct the hash table, those rows will appear identical, and only one will be inserted. Then, when we probe we'll get a hit in the hash table based on the contents of the left equality table, but we'll only see one of the two possible rows from the right conditional table that need to be checked, which means that we may not find a match even if one exists. When we build the hash table we don't need to include every single row in the combined right table (the table including both equality and conditional columns), but we do need to include every _unique_ row so that when probing we test all the possibilities.

The current solution is a stopgap that I put in yesterday just to verify that this fixes the problem. The two better approaches would be to 1) use an equality operator that checks both the equality and conditional tables during building, or 2) to use a multiset (like in inner/left joins) instead of a standard set. Both of those are more involved than the current hack, though, so I was planning to try those today after letting you verify that this brute force approach still fixes the problem.

Reply:
Ah, sorry. I misread the comment as intended for rows of the left table when it's for the right table, and in that context different conditional table contents for the relevant rows is very pertinent!